lib/linked_list.rb

```diff
@@ -3,86 +3,189 @@
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(value)
-      raise NotImplementedError
+      return false if @head.nil? 
+      pointer = @head 
+ 
+      until pointer.next.nil?
+       
+        if pointer.data == value || pointer.next.data == value 
+          return true 
+        else  
+          pointer = pointer.next 
+        end 
+
+      end 
+
+      return false 
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max
-      raise NotImplementedError
+      return nil if @head.nil? 
+      max = @head.data
+      pointer = @head
+
+      while !pointer.nil?
+        if pointer.data > max 
+          max = pointer.data 
+        end 
+        pointer = pointer.next 
+      end 
+      return max 
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_min
-      raise NotImplementedError
+      return nil if @head.nil? 
+
+      min = @head.data
+   
+      pointer = @head
+      while !pointer.nil?
+        if pointer.data < min
+          min = pointer.data 
+        end 
+        pointer = pointer.next 
+      end 
+    
+      return min 
     end
 
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length
-      raise NotImplementedError
+      count = 1 
+      return 0 if @head.nil? 
+      pointer = @head
+      until pointer.next.nil? 
+        count += 1 
+        pointer = pointer.next
+      end 
+      return count 
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(index)
-      raise NotImplementedError
+      node_index = 0 
+      pointer = @head 
+      until pointer.nil? 
+       if node_index == index 
+        return pointer.data 
+       else 
+        node_index += 1 
+       end 
+       pointer = pointer.next 
+      end 
+  
+      return nil 
     end
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def visit
-      raise NotImplementedError
+      pointer = @head 
+      while !pointer.nil? 
+        puts pointer.data 
+        pointer = pointer.next 
+      end 
     end
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
+ 
     def delete(value)

```

Comment:
👍 , but the space complexity is O(1) since you're not making an entirely new list.